### ISSUE TITLE
Jane's ATF with NATO Fighters expansion (1996) 0.82.0 pacing issue

### ISSUE THREAD

**Issue by Grounded0** [dosbox-staging/dosbox-staging#3887]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.82.0 RC1

### What Operating System are you using?

macOS (Intel)

### If Other OS, please describe

macOS 12.6.5

### Relevant hardware info

2017 15" MacBook Pro with Radeon Pro 560

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Jane's ATF with NATO Fighters expansion triggers a weird audio glitch on the intro screen and can be listened in the menus. Regression from 0.81.2. Combat simulation seems to run glitch free probably because it puts an enormous workload on the main thread as these are among the most demanding DOS games ever released:

<img width="1392" alt="Screenshot 2024-08-21 at 12 15 11" src="https://github.com/user-attachments/assets/6d67d2c7-daed-4425-b7f4-b6bd1c3f9b20">

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Put the emulation to balls to the wall setting like thread load 100 %
2. Run Jane's ATF with NATO Fighters expansion
3. Hear the audio glitching on the menus before entering combat simulation

### Your configuration

```ini
https://github.com/user-attachments/files/16689305/dosbox-staging.conf.zip
```
```
ajp@milla:~/Misc/dosbox/c_drive/JANES/ATF $ cat RUN.BAT 
imgmount -u d
imgmount d y:\atf\atf.iso y:\atf\nato.img -t cdrom
config -set "cycles=auto 12000 105%%"
af.exe %1 %2 %3
config -set "cycles=auto 12000 50%%"
if errorlevel = 255 goto exit
if errorlevel = 1 goto TNTerror
goto exit

:TNTerror
echo 

echo 

echo You do not have enough conventional memory to run Advanced Tactical Fighters!
echo Please refer to your Install Guide on making a boot disk that will free
echo a total of 420K of memory.
echo 

goto exit

:exit
ajp@milla:~/Misc/dosbox/c_drive/JANES/ATF $ 
```

### Provide a Log

```text
ajp@milla:~ $ dosbox
date       time         | 
2024-08-21 12:07:22.575 | arguments: dosbox
2024-08-21 12:07:22.575 | Current dir: /Users/ajp
2024-08-21 12:07:22.575 | stderr verbosity: 0
2024-08-21 12:07:22.575 | -----------------------------------
2024-08-21 12:07:22.575 | dosbox-staging version 0.82.0-rc1 (9d55e)
2024-08-21 12:07:22.575 | ---
2024-08-21 12:07:22.575 | LOG: Loguru version 2.1.0 initialised
2024-08-21 12:07:22.577 | CONFIG: Deprecated option 'cycles'
2024-08-21 12:07:22.583 | CONFIG: The 'cycles' setting is deprecated but still accepted; please use the
'cpu_cycles', 'cpu_cycles_protected' and 'cpu_throttle' settings instead as
support will be removed in the future.
2024-08-21 12:07:22.584 | CONFIG: Loaded primary config file '/Users/ajp/Library/Preferences/DOSBox/dosbox-staging.conf'
2024-08-21 12:07:22.584 | LANG: Using internal English language messages
2024-08-21 12:07:22.725 | SDL: version 2.30.6 initialised (cocoa video and coreaudio audio)
2024-08-21 12:07:22.726 | DISPLAY: Using 1280x960 window size in windowed mode on display-0
2024-08-21 12:07:24.081 | OPENGL: Vendor: ATI Technologies Inc.
2024-08-21 12:07:24.081 | OPENGL: Version: 2.1 ATI-4.8.101
2024-08-21 12:07:24.081 | OPENGL: GLSL version: 1.20
2024-08-21 12:07:24.081 | OPENGL: NPOT textures supported
2024-08-21 12:07:24.084 | RENDER: Using shader 'interpolation/sharp'
2024-08-21 12:07:24.084 | VIDEO: Using the DOS video mode's frame rate
2024-08-21 12:07:24.091 | MEMORY: Memory sizes above 31 MB aren't recommended for most DOS games
2024-08-21 12:07:24.115 | MEMORY: Using 8192 DOS memory pages (32 MB) at address: 0x7f884ea30000
2024-08-21 12:07:24.116 | VIDEO: Initialised S3 Trio64 VESA 2.0 compatible with 4 MB of FP DRAM supporting 92 modes
2024-08-21 12:07:24.127 | VOODOO: Initialized with 4 MB of RAM, multithreading, and bilinear filtering
2024-08-21 12:07:24.127 | CAPTURE: Image capturer started
2024-08-21 12:07:24.163 | MIXER: Initialised stereo 48000 Hz audio with 512 sample frame buffer
2024-08-21 12:07:24.163 | MIXER: Master compressor enabled
2024-08-21 12:07:24.190 | FSYNTH: Using SoundFont '/Users/ajp/Library/Preferences/DOSBox/soundfonts/default.sf2'
2024-08-21 12:07:24.190 | FSYNTH: Chorus enabled with 3 voices at level 1.20, 0.30 Hz speed, 8.00 depth, and sine-wave modulation
2024-08-21 12:07:24.190 | FSYNTH: Reverb enabled with a 0.61 room size, 0.23 damping, 0.76 width, and level 0.56
2024-08-21 12:07:24.190 | FSYNTH: Operating at 48000 Hz without resampling
2024-08-21 12:07:24.190 | MIDI: Opened device: fluidsynth
2024-08-21 12:07:24.190 | MPU-401: Running in intelligent mode on ports 330h and 331h
2024-08-21 12:07:24.190 | OPL: Operating at 49716 Hz and downsampling to the output rate
2024-08-21 12:07:24.190 | OPL: Running OPL3 on ports 220h and 388h
2024-08-21 12:07:24.190 | OPL: OPL filter disabled
2024-08-21 12:07:24.190 | DMA: Initialised primary controller
2024-08-21 12:07:24.190 | DMA: Initialised secondary controller
2024-08-21 12:07:24.190 | SB: Operating at 22050 Hz and upsampling to the output rate
2024-08-21 12:07:24.190 | SB16: Modern DAC output filter enabled
2024-08-21 12:07:24.190 | SB16: Setting 'BLASTER' environment variable to 'A220 I7 D1 H5 T6'
2024-08-21 12:07:24.190 | SB16: Running on port 220h, IRQ 7, DMA 1, and high DMA 5
2024-08-21 12:07:24.190 | PCSPEAKER: Operating at 48000 Hz without resampling
2024-08-21 12:07:24.190 | PCSPEAKER: Initialised discrete model
2024-08-21 12:07:24.190 | PCSPEAKER: High-pass filter enabled (18 dB/oct at 120 Hz)
2024-08-21 12:07:24.190 | PCSPEAKER: Low-pass filter enabled (12 dB/oct at 4800 Hz)
2024-08-21 12:07:24.191 | MOUSE: Will move seamlessly: left and right button clicks won't capture the mouse
2024-08-21 12:07:24.191 | MOUSE: Middle button will capture/release the mouse (clicks not sent to the game/program)
2024-08-21 12:07:24.191 | MOUSE (PS/2): Standard, 3 buttons
2024-08-21 12:07:24.193 | DOS: Loaded modern locale for country 0, 'International (English)'
2024-08-21 12:07:24.201 | DOS: Loaded codepage 858 for detected language 'fi'
2024-08-21 12:07:24.201 | SLIRP: Slirp version: 4.7.0
2024-08-21 12:07:24.202 | SLIRP: Successfully initialized
2024-08-21 12:07:24.202 | NE2000: Initialised on port 300h and IRQ 10
2024-08-21 12:07:24.202 | AUTOEXEC: Using autoexec from /Users/ajp/Library/Preferences/DOSBox/dosbox-staging.conf
2024-08-21 12:07:24.203 | MAPPER: No joysticks found
2024-08-21 12:07:24.204 | MAPPER: Loaded 158 key bindings from '/Users/ajp/Library/Preferences/DOSBox/mapper-v2.map'
2024-08-21 12:07:24.207 | MOUNT: Path '~/Misc/dosbox/c_drive' found
2024-08-21 12:07:24.210 | MOUNT: Path '~/Misc/images/iso/dosbox' found
2024-08-21 12:07:24.319 | DOS: Loaded codepage 850
2024-08-21 12:07:24.322 | SDL: Using standard SDI (auto) display refresh rate of 60 Hz
2024-08-21 12:07:24.371 | DISPLAY: VGA 720x400 16-colour text mode 03h at 70.087 Hz throttled VFR, scaled to 2560x1920 pixels with 1:1.35 (20:27) pixel aspect ratio
2024-08-21 12:07:24.371 | OPENGL: Disabled vsync
2024-08-21 12:07:35.210 | IMGMOUNT: Path 'y:\atf\atf.iso' not found, maybe it's a DOS path
2024-08-21 12:07:35.210 | IMGMOUNT: Path 'ATF\ATF.ISO' found on virtual drive Y:
2024-08-21 12:07:35.210 | IMGMOUNT: Path 'y:\atf\nato.img' not found, maybe it's a DOS path
2024-08-21 12:07:35.210 | IMGMOUNT: Path 'ATF\NATO.IMG' found on virtual drive Y:
2024-08-21 12:07:35.210 | CDAUDIO: Operating at 48000 Hz without resampling
2024-08-21 12:07:35.214 | CONFIG: Deprecated option 'cycles'
2024-08-21 12:07:35.214 | CONFIG: The 'cycles' setting is deprecated but still accepted; please use the
'cpu_cycles', 'cpu_cycles_protected' and 'cpu_throttle' settings instead as
support will be removed in the future.
2024-08-21 12:07:35.230 | CPU: Switched to max cycles. Try setting a fixed cycles value if you're getting audio drop-outs or the programs runs too fast.
2024-08-21 12:07:36.451 | SB16: Resetting DSP
2024-08-21 12:07:36.526 | DISPLAY: VESA 640x480 256-colour graphics mode 101h at 70.007 Hz throttled VFR, scaled to 2560x1920 pixels with 1:1 (1:1) pixel aspect ratio
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by Grounded0** [comment]
This is definitely one of those very few games that benefit from a setting like thread load 100 % on its combat simulation.

**Comment by Grounded0** [comment]
I'm also getting same audio glitching at cycles 80 %. Most prominent before entering the combat simulation in Load Ordnance screen where you select your weapons load out. Try switching your weapons load out and the problem should become prominent:

<img width="1392" alt="Screenshot 2024-08-21 at 15 45 39" src="https://github.com/user-attachments/assets/0a71e358-e6d4-4275-a361-b69352a3c990">

**Comment by Grounded0** [comment]
I'm getting same audio glitch at thread load 50 % which maps to Pentium Pro 200 MHz on this system so there's probably something else going on than using too high thread load value. 0.81.2 handles thread load 100 % just fine.

**Comment by johnnovak** [comment]
@interloper98 @weirddan455 Hey mixer gurus, are you going to look into this? Would be nice to fix it for the final if a regression.

**Comment by Grounded0** [comment]
Flagged this as low priority for a reason. Isn't present in the simulation.

**Comment by johnnovak** [comment]
> Labelled this as low priority for a reason. Isn't present in the simulation or the videos.

Thanks for the info. As it's a regression, it would be good to get to the bottom of it, though, but looks like we can then release the final without a fix.

**Comment by interloper98** [comment]
I'm not sure how it should (or shouldn't) sound, but here's what I'm getting:

https://github.com/user-attachments/assets/f076773f-c30c-4684-8af4-00c1f61bd9c4

@Grounded0 , how does that recording compared to what you're hearing?

Maybe try my conf:

```ini
[dosbox]
machine=svga_s3
memsize=128

[cpu]
cpu_cycles_protected = 150000
cpu_throttle = true

[autoexec]
imgmount d cd/ATF_1_03F.iso -t iso
imgmount e cd/NATO_1_00F.img -t iso
mount c .
c:
call run
```

**Comment by Grounded0** [comment]
That's free of the pacing errors I got. I'll try to see if I can encapsulate it in a recording.

**Comment by Grounded0** [comment]
https://anttipeltola.eu/misc/janesatf082audio.wav

**Comment by interloper98** [comment]
@Grounded0, I noticed that the game is somehow very gentle on its CPU use, even when cycles gets up into the Pentium range.

When I configured DOSBox to "consume a percent of the host CPU" (say, `max` or `80%`), the game's gentle behavior meant that the underlying cycle count get massive. Like, over a 1 million cycles per ms... and even then, it hadn't maxed out my CPU.

It looks like DOSBox's auto-cycle logic can get in trouble at these eye popping cycle counts when all of a sudden a burst of much harder works comes down the pipe. Like a wild party run wild.. eventually the police are showing up with machine guns.

The game doesn't need those millions of cycles.. it runs great even with 150K cycles (even less, actually). So by setting the `cpu_cycles_protected = 150000` combined with `cpu_throttle = true`, this prevents DOSBox from throwing a cycle party and getting itself in trouble :-)

**Comment by Grounded0** [comment]
It is one of those games with insane range. Have a mission to destroy adversary airbase and you have AAA fire filling up the skies with your own four ship and adversary fighters engaging in air combat and the game suddenly prefers Pentium III 500 MHz system.

**Comment by interloper98** [comment]
If 150000 still isn't enough for this massive dog fights, then try a bit more; say 300000. Based on what I saw, that should still be a lot safer than using a percent-approach (that will get itself into trouble).

**Comment by johnnovak** [comment]
> Like a party run wild.. eventually the police are showing up with machine guns.

I love these little stories and analogies of yours @interloper98 😆 So much fun and they illustrate the concepts very well 😎

**Comment by interloper98** [comment]
> > Like a party run wild.. eventually the police are showing up with machine guns.
> 
> I love these little stories and analogies of yours @interloper98 😆 So much fun and they illustrate the concepts very well 😎

Thanks @johnnovak ; it's interesting poking around in the CPU and cycle loops! And these new CPU setting offered in 0.82 `cpu_cycles_protected` combined with `cpu_throttle` make it really easy to avoid runaway cycle counts that `max` causes, while still offered the throttle feature.

**Comment by Grounded0** [comment]
Still got pacing errors with default config file with modifications:

``
[dosbox]
machine=svga_s3
memsize=32

[cpu]
cpu_cycles_protected = 300000
cpu_throttle = true

AUTOEXEC.BAT:

@ECHO OFF

rem FreeDOS DOSBox specific AUTOEXEC.BAT file. You may wish to have your                                  
rem Z:\AUTOEXEC.BAT file check for and automatically launch this batch file                               
rem at DOSBox startup.                                                                                    

ECHO Loading C:\AUTOEXEC.BAT...
SET OLDPATH=%PATH%
SET DOSDIR=C:\FDOS
IF "%LANG%" == "" SET LANG=EN
IF "%TZ%" == "" SET TZ=UTC
SET PATH=C:\CMDTOOLS;C:\CMDTOOLS\MXDIR;C:\WINDOWS;C:\APPS\DN2;C:\DOSMID;%dosdir%\BIN;C:\NOUNIVBE;C:\SDD;C\
:\DOSIDLE;Z:\
IF EXIST %dosdir%\LINKS\NUL SET PATH=%path%;%dosdir%\LINKS
SET PATH=c:\bc7\bin;c:\bc7\binb;%PATH%
IF NOT "%OLDPATH%" == "" PATH=%PATH%;%OLDPATH%
SET OLDPATH=

SET NLSPATH=%dosdir%\NLS
SET HELPPATH=%dosdir%\HELP
SET TEMP=%dosdir%\TEMP
SET TMP=%TEMP%
set LIB=c:\bc7\lib;%LIB%
set INCLUDE=c:\bc7\src;%INCLUDE%
set HELPFILES=c:\bc7\help;%HELPFILES%

SET DIRCMD=/P /OGN /Y
SET COPYCMD=/-Y

SET AUTOFILE=%0
SET CFGFILE=C:\CONFIG.SYS
alias cfg=edit %cfgfile%
alias auto=edit %0

set OS_NAME=FreeDOS
set OS_VERSION=1.3

rem MODE 80,43                                                                                            
rem C:\WINDOWS\SMARTDRV.EXE                                                                               
rem univbe -d                                                                                             
rem nounivbe                                                                                              

SET SOUND=C:\SB16
rem SET BLASTER=A220 I7 D1 H5 P330 T6                                                                     
SET MIDI=SYNTH:1 MAP:E
rem C:\SB16\DIAGNOSE /S                                                                                   
rem C:\SB16\MIXERSET /P /Q                                                                                

rem SHARE                                                                                                 

rem dosidle -hlt -fm0                                                                                     

nounivbe

echo Setting PCSPEAKER volume to 50

mixer pcspeaker 50 /noshow

echo Setting FSYNTH volume to 50

mixer fsynth 50 /noshow

AUDIOCFG

rem dn      

``

**Comment by Grounded0** [comment]
I disagree the way DOSBox Staging forcefully runs AUTOEXEC.BAT when its not asked to. Another instance of us trying to be too clever instead of handing 100 % control to the user over this.

**Comment by interloper98** [comment]
@Grounded0 , right. Don't use 300k if it's still causing problems. Test lower peak cycles until it makes it by the ammo load without the audio  hiccups. 

I suspect whatever you find, it will deliver plenty of FPS in-game (I was hitting 60 FPS no problem doing carrier passes at 150k cycles).

> Can I start this somehow in a way that DOSBox Staging isn't trying to be too clever and run AUTOEXEC sections somewhere I didn't ask it to run?
> 
> I'm currently starting with `$ dosbox --conf default.conf`.

I searched the conf for autoexec - looks like Staging has this `[dosbox] autoexec_section`:

```
# autoexec_section: How autoexec sections are handled from multiple config files:
#   join:       Combine them into one big section (legacy behavior; default).
#   overwrite:  Use the last one encountered, like other config settings.
#               Possible values: join, overwrite.
```

**Comment by Grounded0** [comment]
Yeah I was wondering what was going on but of course default.conf turned that into join because my own .conf files have it on overwrite. This was something we put in earlier with kcgen because the default behaviour is braindead and only there for backwards compatibility.

**Comment by johnnovak** [comment]
Dunno, joining autoexec sections can be useful. It's a matter of being aware of the feature and taking advantage of it.

It's not that the autoexec section is a regular config setting. It's not a config setting; it is its own special thing (list of commands). I use the join behaviour in my upcoming DOS pack to implement "add-on" packages on top of a base package.

**Comment by Grounded0** [comment]
Okay spent time converting everything to the new cycles model.

Jane's ATF can handle 450000 cycles with `cpu_throttle=off` without pacing issues on this machine. Then I have to set `cpu_throttle=on` to make the menus bearable.

This game really requires 450000 cycles for smooth operation over large airbases so I will not be dialling this down.

US Navy Fighters including Marine Fighters expansion is now free of pacing issues caused by issuing cpu cycles by thread load value. With those games 350000 seems to be a good value.

**Comment by johnnovak** [comment]
> Okay spent time converting everything to the new cycles model.
> 
> Jane's ATF can handle 450000 cycles with `cpu_throttle=off` without pacing issues on this machine. Then I have to set `cpu_throttle=on` to make the menus sound bearable. This game really requires 450000 cycles for smooth operation over large airbases so I will not be dialling this down.
> 
> US Navy Fighters including Marine Fighters expansion (1995) is now free of pacing issues caused by issuing cpu cycles by thread load value. With those games 350000 seems to be a good value.

Updating the wiki with this info would be worthwhile @Grounded0 as few games require such high cycles settings to run optimally.

**Comment by Grounded0** [comment]
Well there's still the problem of how I can have those menus in Jane's ATF without the pacing issues with the music though still low priority. Previously you could have both and it still worked.

**Comment by Grounded0** [comment]
Seems to be fixed in 0.82.2.
